### ISSUE TITLE
Asyncpg with sqlalchemy cannot perform operations at the same time

### ISSUE THREAD

**Issue by ItsDrike** [sqlalchemy/sqlalchemy#5967]
**Describe the bug**
I'm using the new async engine with asyncpg (postgresql), for a discord bot, that is running with asyncio. I tried to access the database at the same time with 2 same event handlers in different files. This caused asyncpg to raise `asyncpg.exceptions._base.InterfaceError`, which then lead to `sqlalchemy.exc.InterfaceError`. I'm using ORM structure and I tried to perform a read operation like this.

**Expected behavior**
SQLAlchemy should be able to handle running 2 operations at the same time, considering it aims to be asynchronous. Even though it may not happen often, there are cases when the first operation is still running and another request has already been made. Perhaps that operation should wait (asynchronously), and only execute once the first operation is finished, to prevent this error.

**To Reproduce**
If you want a complete reproduction, you can checkout this commit in my bot repository: https://github.com/Codin-Nerds/Neutron-Bot/commit/3aeabbed04bec342d970497bd5751f940482f196 and run the bot (with PostgreSQL hosted, or in docker, follow README for more details).

But simplified: when 2 operations are made asynchronously at the same time, this error occurs

<details>
 <summary><b>Error</b></summary>

```
Ignoring exception in on_member_update
Traceback (most recent call last):
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 530, in _start_transaction
    await self._transaction.start()
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/asyncpg/transaction.py", line 138, in start
    await self._connection.execute(query)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/asyncpg/connection.py", line 295, in execute
    return await self._protocol.query(query, timeout)
  File "asyncpg/protocol/protocol.pyx", line 301, in query
  File "asyncpg/protocol/protocol.pyx", line 664, in asyncpg.protocol.protocol.BaseProtocol._check_state
asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1628, in _execute_context
    self.dialect.do_execute(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 663, in do_execute
    cursor.execute(statement, parameters)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 350, in execute
    self._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 290, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 513, in _handle_exception
    raise error
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 346, in execute
    self._adapt_connection.await_(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 41, in await_only
    return current.driver.switch(awaitable)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 86, in greenlet_spawn
    value = await result
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 312, in _prepare_and_execute
    await self._adapt_connection._start_transaction()
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 532, in _start_transaction
    self._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 509, in _handle_exception
    raise translated_error from error
sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.InterfaceError: <class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/discord/client.py", line 343, in _run_event
    await coro(*args, **kwargs)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/bot/cogs/logging/member_log.py", line 84, in on_member_update
    member_log_id = await LogChannels.get_log_channel(self.bot.db_session, "member_log", member_after.guild)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/bot/database/log_channels.py", line 76, in get_log_channel
    log_channels = await cls.get_log_channels(session, guild)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/bot/database/log_channels.py", line 64, in get_log_channels
    row = await session.run_sync(lambda session: session.query(cls).filter_by(guild=guild).one())
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/ext/asyncio/session.py", line 120, in run_sync
    return await greenlet_spawn(fn, self.sync_session, *arg, **kw)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 91, in greenlet_spawn
    result = context.throw(*sys.exc_info())
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/bot/database/log_channels.py", line 64, in <lambda>
    row = await session.run_sync(lambda session: session.query(cls).filter_by(guild=guild).one())
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2695, in one
    return self._iter().one()
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 2733, in _iter
    result = self.session.execute(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1530, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1443, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 310, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1312, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1671, in _execute_context
    self._handle_dbapi_exception(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1852, in _handle_dbapi_exception
    util.raise_(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 180, in raise_
    raise exception
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1628, in _execute_context
    self.dialect.do_execute(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 663, in do_execute
    cursor.execute(statement, parameters)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 350, in execute
    self._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 290, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 513, in _handle_exception
    raise error
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 346, in execute
    self._adapt_connection.await_(
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 41, in await_only
    return current.driver.switch(awaitable)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 86, in greenlet_spawn
    value = await result
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 312, in _prepare_and_execute
    await self._adapt_connection._start_transaction()
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 532, in _start_transaction
    self._handle_exception(error)
  File "/home/itsdrike/Programming/Python/GitHub/Codin-Nerds/Neutron-Bot/.venv/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 509, in _handle_exception
    raise translated_error from error
sqlalchemy.exc.InterfaceError: (sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) <class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress
[SQL: SELECT log_channels.guild AS log_channels_guild, log_channels.server_log AS log_channels_server_log, log_channels.mod_log AS log_channels_mod_log, log_channels.message_log AS log_channels_message_log, log_channels.member_log AS log_channels_member_log, log_channels.join_log AS log_channels_join_log, log_channels.voice_log AS log_channels_voice_log 
FROM log_channels 
WHERE log_channels.guild = ?]
[parameters: ('608625191068631040',)]
(Background on this error at: http://sqlalche.me/e/14/rvf5)
```
</details>

**Versions.**
 - OS: Arch Linux (linux 5.10.16.arch1-1)
 - Python: 3.8
 - SQLAlchemy: 1.4.0b1
 - Database: PostgreSQL 13.2-1
 - DBAPI: I'm not familiar with what this is

**Comment by CaselIT** [comment]
Hi,

Regarding your issue if I understood correctly it seems that you are trying to re-use the same connection without first exhausting the first result set.
If that's the case I'm not sure what sqlalchemy could to help. The two operations should probably use two different connections if it's not possible to complete the first one before the second one is run.
Without a runnable reproduction its difficult to guess.

In any case, could you try with the b3 to see if there is the same problem? IIRC there were some things that were changed regarding asyncpg.

**Comment by ItsDrike** [comment]
> Without a runnable reproduction its difficult to guess

You can reproduce the issue by running the bot from my repository, it is fully open. You'll then want to simply trigger that handler by giving anyone any role. The commit in which this happened is marked in the question. But I can see that not many people would be willing to set the whole bot up just to try and test this.

> b3 to see if there is the same problem

Yes, the issue does happen with b3 version too, there isn't any difference.

> Regarding your issue if I understood correctly it seems that you are trying to re-use the same connection without first exhausting the first result set.

Well, it is running 2 listeners at relatively the same time, when 1 database read is io operation, so it likely switches to the other listener and tries to obtain the same thing from my database again, without the first being used in any way yet.

**Comment by zzzeek** [comment]
SQLAlchemy has no inherent limitations on concurrency.  If you are seeing asyncpg InterfaceError, assuming this is a simple matter of running multiple concurrent operations on a single connection, those are limitations in asyncpg's prepared statement interface for which I am unaware of any mitigation.   SQLAlchemy must use the prepared statement interface because it requires the names and datatypes of the columns that would be returned which are delivered by the "attributes" collection, and asyncpg does not seem to provide another way to get this information.

 See the demonstration below.  As asyncpg does not provide the standard DBAPI cursor interface, they do not seem to provide thread safety at the connection level.     

Drivers like psycopg2 do in fact provide connection-level thread safety as described at https://www.psycopg.org/docs/usage.html#thread-and-process-safety by encapsulating the state of operations within individual cursors, so if you used psycopg2 then this would be possible, but not with the asyncio interface.

The other way to resolve is to use a distinct connection per async task.   the below example illustrates this as well.

Feel free to the asyncpg [issue tracker](https://github.com/MagicStack/asyncpg/issues) to see if they have any means by which they might allow multiple concurrent prepared statements.    

```python
import asyncio
import asyncpg


async def run_asyncpg_statement(connection, argument):
    # uncomment to fix
    # connection = await asyncpg.connect(
    #     user="scott", password="tiger", host="pg12", database="test"
    # )

    prepared_stmt = await connection.prepare("SELECT $1::int AS some_data")

    result = await prepared_stmt.fetch(argument)
    print(f"result: {result}")


async def main():

    connection = await asyncpg.connect(
        user="scott", password="tiger", host="pg12", database="test"
    )

    await asyncio.gather(
        *[run_asyncpg_statement(connection, i) for i in range(10)]
    )


asyncio.run(main())

```

the output includes multiple errors to the effect of "asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress".

using connection per task, the output has no errors:

```
 python test3.py 
result: [<Record some_data=0>]
result: [<Record some_data=4>]
result: [<Record some_data=6>]
result: [<Record some_data=5>]
result: [<Record some_data=1>]
result: [<Record some_data=9>]
result: [<Record some_data=7>]
result: [<Record some_data=8>]
result: [<Record some_data=3>]
result: [<Record some_data=2>]

```

**Comment by CaselIT** [comment]
> > Without a runnable reproduction its difficult to guess
> 
> You can reproduce the issue by running the bot from my repository, it is fully open. You'll then want to simply trigger that handler by giving anyone any role. The commit in which this happened is marked in the question. But I can see that not many people would be willing to set the whole bot up just to try and test this.

By runnable reproduction I meant something like the example in mikes answer.
 
Returning to the issue at hand, like mike suggested, try use a different connection (aka session if using the orm) for each listener.

**Comment by zzzeek** [comment]
even using a lock doesnt seem to work, which would be one thing we could maybe do...but it doesn't work anyway:
```python

lock = asyncio.Lock()

async def run_asyncpg_statement(connection, argument):

    async with lock:
        prepared_stmt = await connection.prepare("SELECT $1::int AS some_data")

        result = await prepared_stmt.fetch(argument)
    print(f"result: {result}")

```

**Comment by ItsDrike** [comment]
> even using a lock doesnt seem to work, which would be one thing we could maybe do...but it doesn't work anyway:
> 
> ```python
> lock = asyncio.Lock()
> 
> async def run_asyncpg_statement(connection, argument):
> 
>     async with lock:
>         prepared_stmt = await connection.prepare("SELECT $1::int AS some_data")
> 
>         result = await prepared_stmt.fetch(argument)
>     print(f"result: {result}")
> ```

This was my course of thinking too, and I was trying the exact same thing

> Returning to the issue at hand, like mike suggested, try use a different connection (aka session if using the orm) for each listener.

This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like `session.wait_until_free()` which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.

Adding something like this would be helpful for all libraries which aren't thread-safe, just like in the asyncpg's case, and so it seems like a good addition if it's possible to implement it.

**Comment by zzzeek** [comment]
I'm not really understanding why asyncpg doesn't work with the lock in any case, and someone should figure out what's wrong with my example, because i dont see how there would be an issue with multiple async tasks using the same connection as long as they run and complete their full query before going out to any other worker.

if we can't mutex around the asyncpg execution then i dont see how any other strategy would work, becasue they will all have to use a lock around the operation in some way.   like i think you should put a mutex around your own use of the session to solve this problem, but that's no different than the way I'm using a mutex above. 

Also I'm -1 on inventing concurrency primitives in SQLAlchemy, these arleady exist and are out of scope for a database abstraction library, and this is definitely an asyncio only issue as in the threading world people are not sharing sessions between threads.

**Comment by zzzeek** [comment]
i dont really understand how this fails, even using separate connections, it fails.     but if i just use the lock without running asyncpg in the middle, then it's fine.   Someone needs to explain to me what is happening here.

```python
import asyncio
import asyncpg


lock = asyncio.Lock()

async def run_asyncpg_statement(connection, argument):
    # uncomment to fix
    connection = await asyncpg.connect(
        user="scott", password="tiger", host="pg12", database="test"
    )

    async with lock:
        prepared_stmt = await connection.prepare("SELECT $1::int AS some_data")

        result = await prepared_stmt.fetch(argument)
        print(f"result: {result}")


async def main():

    connection = await asyncpg.connect(
        user="scott", password="tiger", host="pg12", database="test"
    )

    await asyncio.gather(
        *[run_asyncpg_statement(connection, i) for i in range(10)]
    )


asyncio.run(main())
```

```
$ python test3.py 
Traceback (most recent call last):
  File "test3.py", line 31, in <module>
    asyncio.run(main())
  File "/opt/python-3.8.3/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/opt/python-3.8.3/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "test3.py", line 26, in main
    await asyncio.gather(
  File "test3.py", line 13, in run_asyncpg_statement
    async with lock:
  File "/opt/python-3.8.3/lib/python3.8/asyncio/locks.py", line 97, in __aenter__
    await self.acquire()
  File "/opt/python-3.8.3/lib/python3.8/asyncio/locks.py", line 203, in acquire
    await fut
RuntimeError: Task <Task pending name='Task-4' coro=<run_asyncpg_statement() running at test3.py:13> cb=[gather.<locals>._done_callback() at /opt/python-3.8.3/lib/python3.8/asyncio/tasks.py:751]> got Future <Future pending> attached to a different loop
Task was destroyed but it is pending!
task: <Task pending name='Task-23' coro=<Connection._cancel() done, defined at /home/classic/.venv3/lib/python3.8/site-packages/asyncpg/connection.py:1211> wait_for=<Future pending cb=[BaseSelectorEventLoop._sock_write_done(19)(), <TaskWakeupMethWrapper object at 0x7f2f6aa47040>()]>>

```

**Comment by CaselIT** [comment]
Don't create the lock at module level mike and it should fix that exception.

The different loop was one issue that we fixed in one of the beta

**Comment by zzzeek** [comment]
ohh create it in main, right

**Comment by zzzeek** [comment]
wow that's annoying.  OK so anyway, if you want to spin off another Session operation, put it inside of a Lock.    SQLAlchemy itself maybe could put this lock inside the asyncpg dialect but we don't do that for any other DBAPI and I would rather not start.     I dont understand what "wait_until_free()" would do (besides give the caller ...a lock!  that they should just use themselves).

**Comment by CaselIT** [comment]
> This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like session.wait_until_free() which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.

I'm not sure exactly how the discord bot usually calls the events, but if you have a bunch of events that are fired at different times, that sound similar to a classic web server, and it that case you want to have a separate session for each request, event in this case.
This way there could be many concurrent events and everything would work correctly since each one uses its own local session.

Let me know it the analogy does not hold for your use case

**Comment by ItsDrike** [comment]
> > This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like session.wait_until_free() which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.
> 
> I'm not sure exactly how the discord bot usually calls the events, but if you have a bunch of events that are fired at different times, that sound similar to a classic web server, and it that case you want to have a separate session for each request, event in this case.
> This way there could be many concurrent events and everything would work correctly since each one uses its own local session.
> 
> Let me know it the analogy does not hold for your use case

Well, I suppose it is a bit similar to a web server, but there are also some commands, which do communicate with the database, these commands can be triggered from any guild that the bot will be on (which may be a lot), at any given time. I don't feel like it would make sense to also make a whole new session for this occasion too. 

But if that is the usual use case, I suppose I could make each call to the database create its own new session then, to avoid repetition, and in most of the cases, commands or events will only be referring to the database once anyway. I'm not sure how efficient it would be to actually do this every time, in the projects I worked on in the past, I usually kept a single session only.

**Comment by zzzeek** [comment]
in the case of the asyncio drivers we are having to emulate our own DBAPI and cursor, and the proposal here is that the Lock is local to that connection, meaning, it does not prevent any concurrency that is would be allowed in any case.   it looks like aiomysql already gives us the result in one await call.  for asyncpg withotu server side cursors, it's two.

but note that the ORM Session and all that is still not concurrent-safe, this is just the "connection.execute()" part of it, which does not even accommodate for loading of ORM objects, where conflicts could still occur.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**mutex asyncpg connection state changes** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2603

**Comment by zzzeek** [comment]
> > This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like session.wait_until_free() which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.
> 
> I'm not sure exactly how the discord bot usually calls the events, but if you have a bunch of events that are fired at different times, that sound similar to a classic web server, and it that case you want to have a separate session for each request, event in this case.
> This way there could be many concurrent events and everything would work correctly since each one uses its own local session.

in async programming, it is a common need that people want to be able to spin up several IO bound operations and wait on them concurrently, even in the scope of a "request" .  the classic use case, which is what actually convinced me an asyncio DB interface is even useful, is that the request needs to pull information from several web services but there's no relationship between their responses.   the request can stack those calls so they all wait on IO concurrently.   I can see wanting to put some database queries in that same space.

this is not at all how we'd do things in the threaded world, there would always have to be some kind of thread pool for this kind of thing and these are uncommon in web-request oriented systems.

**Comment by CaselIT** [comment]
> 
> 
> > > This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like session.wait_until_free() which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.
> > 
> > 
> > I'm not sure exactly how the discord bot usually calls the events, but if you have a bunch of events that are fired at different times, that sound similar to a classic web server, and it that case you want to have a separate session for each request, event in this case.
> > This way there could be many concurrent events and everything would work correctly since each one uses its own local session.
> 
> in async programming, it is a common need that people want to be able to spin up several IO bound operations and wait on them concurrently, even in the scope of a "request" . the classic use case, which is what actually convinced me an asyncio DB interface is even useful, is that the request needs to pull information from several web services but there's no relationship between their responses. the request can stack those calls so they all wait on IO concurrently. I can see wanting to put some database queries in that same space.
> 
> this is not at all how we'd do things in the threaded world, there would always have to be some kind of thread pool for this kind of thing and these are uncommon in web-request oriented systems.

Sure, but using the lock will serialize them, avoiding making the io operation concurrently, so this pattern does not work correctly using a single connection, since you are not achieving concurrent io. It would be like using a threadpool with a single thread.

I'm not saying that's wrong to use the lock, since not doing so leads to errors and other issues, just that it should be mentioned in the asyncpg dialect. Something like:

> each connection (session in the orm world) supports running one operation at a time, guarded by a lock. To achieve concurrent io operations use different connection/session.

**Comment by CaselIT** [comment]
> 
> 
> > > This would work, but it would be very weird to have to create a whole new session just for this exact occasion, that's why I asked about something like session.wait_until_free() which wouldn't solve the underlying issue directly, since it's asyncpg's issue, but it would provide a way to avoid the issue since just simple locking isn't an option.
> > 
> > 
> > I'm not sure exactly how the discord bot usually calls the events, but if you have a bunch of events that are fired at different times, that sound similar to a classic web server, and it that case you want to have a separate session for each request, event in this case.
> > This way there could be many concurrent events and everything would work correctly since each one uses its own local session.
> > Let me know it the analogy does not hold for your use case
> 
> Well, I suppose it is a bit similar to a web server, but there are also some commands, which do communicate with the database, these commands can be triggered from any guild that the bot will be on (which may be a lot), at any given time. I don't feel like it would make sense to also make a whole new session for this occasion too.
> 
> But if that is the usual use case, I suppose I could make each call to the database create its own new session then, to avoid repetition, and in most of the cases, commands or events will only be referring to the database once anyway. I'm not sure how efficient it would be to actually do this every time, in the projects I worked on in the past, I usually kept a single session only.

I still think that using a separate connection for each event invocation would be the preferred solution since it would avoid any concurrency problem.
But if you think it would best to avoid it, if two events of the same time cannot happen at the same time, you could have a connection for each listener, this way two listeners can be called concurrently without problems.

**Comment by zzzeek** [comment]
once I realized we are also opening the transaction and all that it seems like we should probably mutex here so taht people aren't getting complex failure scenarios.   

also re: my aiomysql opinion, I realized that if they aren't mutexing internally then we still could provide a mutex to prevent complex failure modes when people eventually do this there.

**Comment by zzzeek** [comment]
aiomysql is not automatically doing begin().   how are our tests passing?  hmm

**Comment by zzzeek** [comment]
ah, mysql autocommit is false in the connection.  server does it.      

but yes, this can get confused:

```
    async def query(self, sql, unbuffered=False):
        # logger.debug("DEBUG: sending query: %s", _convert_to_str(sql))
        if isinstance(sql, str):
            sql = sql.encode(self.encoding, 'surrogateescape')
        await self._execute_command(COMMAND.COM_QUERY, sql)
        await self._read_query_result(unbuffered=unbuffered)
        return self._affected_rows

```

**Comment by CaselIT** [comment]
I guess this should also be guarded, since it could happen than two coroutine that use the same connection both call `_execute_command` before either has a chance to call `_read_query_result`. So in that sense also aiomysql is not concurrent save

But by this reasoning, shouldn't we also verify that each normal (ei non async) dbapi is also threadsafe and use a mutex if not? Are they normally?

Is concurrent execution not something that is usually handled by async libraries?

**Comment by acnebs** [comment]
Is the solution here not to use asyncpg's built-in [Connection Pools](https://magicstack.github.io/asyncpg/current/usage.html#connection-pools)?

As mentioned in this issue: https://github.com/MagicStack/asyncpg/issues/695#issuecomment-766142868

**Comment by zzzeek** [comment]
SQLAlchemy has its own connection pool, so when you use engine.connect(), you are using that pool.   Asyncpg's connection pool does not make its own "connection" concurrency-safe, it requires that you check out connections explicitly with aquire().  same as engine.connect().

**Comment by acnebs** [comment]
Gotcha. But I guess my follow-up question is then: why not just use asyncpg's pooling under-the-hood, to allow for what looks like re-using a connection?

As mentioned in the linked thread, multiplexing is not supported by PostgreSQL's protocol anyway, so not sure (asyncio or no) you'd ever *actually* have concurrency on the *same* connection.

```py
import asyncio
import asyncpg


POSTGRES_DSN = "postgresql://connor@localhost/test"


async def run_asyncpg_statement(pool, argument):
    async with pool.acquire() as connection:
        prepared_stmt = await connection.prepare("SELECT $1::int AS some_data")
        result = await prepared_stmt.fetch(argument)
    print(f"result: {result}")


async def main():
    pool = await asyncpg.create_pool(POSTGRES_DSN)
    await asyncio.gather(*[run_asyncpg_statement(pool, i) for i in range(10)])


asyncio.run(main())
```

**Comment by zzzeek** [comment]
> Gotcha. But I guess my follow-up question is then: why not just use asyncpg's pooling under-the-hood, to allow for what looks like re-using a connection?

because....people should be explicit about patterns like this.   if you want to use two connections, you should use two connections, and additionally, you should manage the transactional boundaries of that connection.     what would happen if I did .execute() on two concurrent awaitables then did connection.commit()?  which transaction commits?   Do i hold onto every concurent connection that's been used until the user states what they want to do with each one?

SQLAlchemy is not unaware of this pattern, in our first releases, we had no "connection", we had just the "engine", and "engine.execute".  it used a different connection for every execution.   then one day someone wanted to use a transaction that wasn't linked to a thread local....oops.    the rest is history, "engine.execute" was a bad design decision and finally in 2.0 it will be completely removed.

**Comment by elprans** [comment]
> it requires that you check out connections explicitly with aquire(). same as engine.connect().

FWIW, `Pool` exposes the `fetch*()` and `execute()` methods directly.  IOW, you can treat an `asyncpg.Pool` instance as a concurrency-safe "connection" that you can pass to your `asyncio.gather`, and this is what we recommend people do in application code.  The real reason why `asyncpg.Pool` isn't useful for SQLAlchemy is because you're relying on explicit prepared statements, and those are connection-bound.  Transactions too.

**Comment by zzzeek** [comment]
we just put an asyncio.Lock() around the prepare/execute() phases and this solves the problem nicely.  I just had to come around to doing that.   trying to run the same blocking connection in multiple threads without mutexes is not something anyone thought they should be doing for 15 years, but within the first few months of asyncio, probably three or four people already wanted to do it with a non-blocking connection, so, bit of a cultural difference to accommodate there.

**Comment by elprans** [comment]
> we just put an asyncio.Lock() around the prepare/execute() phases and this solves the problem nicely

Yeah, that's a good library-level solution if you want hassle-free concurrency safety by default.  

The reason why we don't do it in asyncpg is that we want the downstream to be aware and explicit in their choice to implicitly serialize concurrent operations on a connection. Otherwise, users may get a false sense of I/O concurrency where there isn't one.  Also, that "cannot perform operation" asyncpg error message could use some love, I don't think it clearly explains what it means.

**Comment by CaselIT** [comment]
IIRC we need to use prepare + fetch/execute because sqlalchemy needs the information about the types of the select using the attributes: 
https://github.com/sqlalchemy/sqlalchemy/blob/85c843bcc3ea1ca16be06fe85b03d99c4e519f3d/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L392-L406

Is there an alternative to prepare+execute that allows getting the attributes? I guess it would not change the situation wrt the use of the lock since asyncpg does not allow concurrent execution in the same connection in any case.

**Comment by elprans** [comment]
> Is there an alternative to prepare+execute that allows getting the attributes?

No. Taking an explicit prepared statement object is the appropriate way to introspect statements.  The connection-level `fetch` methods are basically just simple wrappers around `prepare()+execute()`.

> I guess it would not change the situation 

Correct.  Every top-level I/O operation on a connection object (`fetch`, `prepare` etc) must be serialized if you cannot avoid concurrent use of connection.

**Comment by CaselIT** [comment]
Ok, thanks!

**Comment by stijnli** [comment]
We addressed it with the code below. 
```
class LockedAsyncSession(AsyncSession):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._lock = asyncio.Lock()

        # Dynamically wrap all coroutine methods with a lock
        for attr_name in dir(self):
            attr = getattr(self, attr_name)
            if inspect.iscoroutinefunction(attr):
                setattr(self, attr_name, self._wrap_with_lock(attr))

    def _wrap_with_lock(self, method):
        async def locked_method(*args, **kwargs):
            async with self._lock:
                return await method(*args, **kwargs)
        return locked_method
```

**Comment by CaselIT** [comment]
It doesn't make much sense, since you aren't using connection concurrently if you have a look.
